### ISSUE TITLE
trouble when running 'rails server' in WSL

### ISSUE THREAD

**Issue by SemperEtAnte** [jruby/jruby#6129]
I'm trying to run 'rails server' on jruby 9.2.11.0 with ruby 2.5.7, but always gives such an error. JRuby, rails and all gems installed in Ubuntu subsystem in Windows 10. rails server with only ruby 2.7.0 works fine.

**Erros trace:**
```
=> Booting Puma
=> Rails 6.0.2.1 application starting in development
=> Run `rails server --help` for more startup options
Exiting
Errno::EPERM: Operation not permitted - 48
                              initialize at org/jruby/RubyIO.java:999
                                     new at org/jruby/RubyIO.java:876
                              initialize at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/rb-inotify-0.10.1/lib/rb-inotify/notifier.rb:64
                              _configure at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/adapter/linux.rb:31
                               configure at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/adapter/base.rb:45
                                    each at org/jruby/RubyArray.java:1809
                               configure at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/adapter/base.rb:40
                                   start at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/adapter/base.rb:63
                                   start at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/forwardable.rb:229
                                Listener at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/listener.rb:68
                           instance_eval at org/jruby/RubyBasicObject.java:2615
                                    call at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/fsm.rb:121
              transition_with_callbacks! at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/fsm.rb:91
                              transition at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/fsm.rb:57
                                   start at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/listen-3.1.5/lib/listen/listener.rb:91
                                   boot! at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/evented_file_update_checker.rb:113
                              initialize at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/evented_file_update_checker.rb:68
                         initialize_i18n at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/i18n_railtie.rb:64
                                 Railtie at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/i18n_railtie.rb:22
                            execute_hook at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:69
                  with_execution_control at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:62
                            execute_hook at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:67
                          run_load_hooks at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:52
                                    each at org/jruby/RubyArray.java:1809
                          run_load_hooks at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:51
                                Finisher at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/application/finisher.rb:129
                           instance_exec at org/jruby/RubyBasicObject.java:2687
                                     run at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/initializable.rb:32
                        run_initializers at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/initializable.rb:61
                              tsort_each at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/tsort.rb:228
       each_strongly_connected_component at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/tsort.rb:350
  each_strongly_connected_component_from at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/tsort.rb:431
       each_strongly_connected_component at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/tsort.rb:349
                                    each at org/jruby/RubyArray.java:1809
                                    call at org/jruby/RubyMethod.java:131
       each_strongly_connected_component at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/tsort.rb:347
                              tsort_each at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/tsort.rb:226
                              tsort_each at /home/semperante/.rvm/rubies/jruby-9.2.11.0/lib/ruby/stdlib/tsort.rb:205
                        run_initializers at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/initializable.rb:60
                             initialize! at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/application.rb:363
                                  <main> at /mnt/d/programs/rubyonrails/ruby_java/config/environment.rb:5
                                 require at org/jruby/RubyKernel.java:974
                        require_relative at org/jruby/RubyKernel.java:1002
                         new_from_string at config.ru:3
                                    eval at org/jruby/RubyKernel.java:1048
                         new_from_string at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/rack-2.2.2/lib/rack/builder.rb:116
                               load_file at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/rack-2.2.2/lib/rack/builder.rb:105
                              parse_file at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/rack-2.2.2/lib/rack/builder.rb:66
       build_app_and_options_from_config at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/rack-2.2.2/lib/rack/server.rb:349
                                     app at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/rack-2.2.2/lib/rack/server.rb:249
                             wrapped_app at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/rack-2.2.2/lib/rack/server.rb:422
                           log_to_stdout at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/commands/server/server_command.rb:77
                                   start at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/commands/server/server_command.rb:37
                                 perform at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/commands/server/server_command.rb:147
                                     tap at org/jruby/RubyKernel.java:1897
                                 perform at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/commands/server/server_command.rb:138
                                     run at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/thor-1.0.1/lib/thor/command.rb:27
                          invoke_command at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/thor-1.0.1/lib/thor/invocation.rb:127
                                dispatch at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/thor-1.0.1/lib/thor.rb:392
                                 perform at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/command/base.rb:69
                                  invoke at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/command.rb:46
                                  <main> at /home/semperante/.rvm/gems/jruby-9.2.11.0/gems/railties-6.0.2.1/lib/rails/commands.rb:18
                                 require at org/jruby/RubyKernel.java:974
                                  <main> at bin/rails:4
```

**The GEMFILE**
```
source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby '2.5.7', :engine => 'jruby', :engine_version => '9.2.11.0'

# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
gem 'rails', '~> 6.0.2', '>= 6.0.2.1'
# Use postgresql as the database for Active Record
gem "activerecord", " 6.0.2.1"
gem 'activerecord-jdbcpostgresql-adapter', :platform => :jruby
# Use Puma as the app server
gem 'puma', '= 4.3.3'

# Reduces boot times through caching; required in config/boot.rb
gem 'bootsnap', '>= 1.4.2', require: false

# Use Rack CORS for handling Cross-Origin Resource Sharing (CORS), making cross-origin AJAX possible
gem 'rack-cors'
group :development, :test do
  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
end

group :development do
  gem 'listen', '>= 3.0.5', '< 3.2'
  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
  gem 'spring'
  gem 'spring-watcher-listen', '~> 2.0.0'
end

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]

gem 'jwt'
gem 'bcrypt', '~> 3.1.7'

gem 'rails_admin'
gem 'sprockets', '~>3.0'
gem 'rspec-rails'
gem 'rswag-specs'
gem 'rswag-api'
gem 'rswag-ui'
gem 'multipart-post'
gem 'rest-client'
```

**Comment by SemperEtAnte** [comment]
**fixed** with replacing gem `rb-inotify-0.10.1` with `rb-inotify-0.9.5`

**Comment by headius** [comment]
I'm going to reopen this because downgrading is only a workaround; we need to make sure the latest versions of `rb-inotify` are working properly on JRuby.

**Comment by headius** [comment]
@SemperEtAnte Am I correct in thinking you are running under Windows Subsystem for Linux?

This is not a scenario we have tested heavily, but I suspect there's some compatibility gap we're handling poorly.

**Comment by kevin-klein** [comment]
I can confirm that i hit this under WSL.

**Comment by headius** [comment]
@kevin-klein Thank you for that, we will try to get it working better for 9.2.14.

**Comment by ahorek** [comment]
here's a simplified test case
```ruby
require 'ffi'

module Native
  extend FFI::Library
  ffi_lib FFI::Library::LIBC

  attach_function :inotify_init, [], :int
end

file_descriptor = Native.inotify_init
@handle = IO.new(file_descriptor)       # => Errno::EPERM (Operation not permitted - 13)
```

**Comment by enebo** [comment]
Confirmed this is still broken on 9.2 branch

**Comment by headius** [comment]
I spent a little time poking at this but I am still not sure why we see this EPERM. The fstat call used to determine if the inotify file descriptor "5" is still open does seem to bottom out in the correct libc call, but for this fd it returns -1 and the errno is 1 (EPERM). Other native IO operations appear to be working fine, but this descriptor triggers an error.

It is also not yet clear to me what might have changed in rb-inotify to cause this, unless it has been broken for quite some time.

We will have to dig into this further in 9.2.14 or 9.3, whichever comes first.

**Comment by headius** [comment]
I did find this WSL issue related to inotify and EPERM that does not seem to have been fully resolved.

https://github.com/microsoft/WSL/issues/216

If the code provided by ahorek above is indeed working on CRuby 2.7.0 then it may have some fallback logic we lack.

**Comment by headius** [comment]
After chatting with @ahorek it is clear this is a combination of WSL not fully supporting inotify and us doing fstat in places that CRuby does not.

The call that fails in @ahorek example is an fstat we use to determine if the given descriptor is still open, in order to safely attach our wrapper ChannelFD to it. The inotify file descriptor is not stat'able and produces this EPERM.

If the example is modified to stat the successful resulting IO object on CRuby, it raises a similar error:

```
require 'ffi'

module Native
  extend FFI::Library
  ffi_lib FFI::Library::LIBC

  attach_function :inotify_init, [], :int
end

file_descriptor = Native.inotify_init
@handle = IO.new(file_descriptor)
@handle.stat
```
```
$ ruby blah.rb
Traceback (most recent call last):
        1: from blah.rb:12:in `<main>'
blah.rb:12:in `stat': Operation not permitted (Errno::EPERM)
```

In order to avoid raising this error we will need to look at alternative ways of checking if the given file descriptor is open, or else reevaluate how we manage this fileno to ChannelFD mapping.

**Comment by ahorek** [comment]
<details>
  <summary>rb-inotify-0.10.1</summary>
  <pre>
.FFFFFFFFFFFFF

Failures:

  1) INotify::Notifier instance stops
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 19
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  2) INotify::Notifier instance process gets events
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 22
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  3) INotify::Notifier instance process gets simultaneous events
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 25
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  4) INotify::Notifier instance process separates events between watches
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 28
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  5) INotify::Notifier instance run processes repeatedly until stopped
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 31
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  6) INotify::Notifier instance run can be stopped from within a callback
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 34
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  7) INotify::Notifier instance run can be stopped before it starts processing
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 37
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  8) INotify::Notifier instance fd returns an integer
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 40
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  9) INotify::Notifier instance to_io returns a ruby IO
     Failure/Error: @handle = IO.new(fd)

     Errno::EPERM:
       Operation not permitted - 43
     # ./lib/rb-inotify/notifier.rb:65:in `initialize'
     # ./spec/notifier_spec.rb:10:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  10) INotify::Notifier instance to_io matches the fd
      Failure/Error: @handle = IO.new(fd)

      Errno::EPERM:
        Operation not permitted - 46
      # ./lib/rb-inotify/notifier.rb:65:in `initialize'
      # ./spec/notifier_spec.rb:10:in `block in <main>'
      # ./spec/notifier_spec.rb:8:in `block in <main>'

  11) INotify::Notifier instance to_io caches its result
      Failure/Error: @handle = IO.new(fd)

      Errno::EPERM:
        Operation not permitted - 49
      # ./lib/rb-inotify/notifier.rb:65:in `initialize'
      # ./spec/notifier_spec.rb:10:in `block in <main>'
      # ./spec/notifier_spec.rb:8:in `block in <main>'

  12) INotify::Notifier instance to_io is selectable
      Failure/Error: @handle = IO.new(fd)

      Errno::EPERM:
        Operation not permitted - 52
      # ./lib/rb-inotify/notifier.rb:65:in `initialize'
      # ./spec/notifier_spec.rb:10:in `block in <main>'
      # ./spec/notifier_spec.rb:8:in `block in <main>'

  13) INotify::Notifier mixed instances doesn't tangle fds
      Failure/Error: @handle = IO.new(fd)

      Errno::EPERM:
        Operation not permitted - 55
      # ./lib/rb-inotify/notifier.rb:65:in `initialize'
      # ./spec/notifier_spec.rb:172:in `block in <main>'
      # ./spec/notifier_spec.rb:172:in `block in <main>'

Finished in 0.36382 seconds (files took 8.62 seconds to load)
14 examples, 13 failures
  </pre>
</details>

<details>
  <summary>rb-inotify-0.9.5</summary>
  <pre>
.....F.F.FFFFF

Failures:

  1) INotify::Notifier instance run processes repeatedly until stopped
     Failure/Error: expect(events.map(&:name)).to match_array(%w(one.txt two.txt))

       expected collection contained:  ["one.txt", "two.txt"]
       actual collection contained:    ["one.txt", "three.txt", "two.txt"]
       the extra elements were:        ["three.txt"]
     # ./spec/notifier_spec.rb:99:in `block in <main>'
     # ./spec/notifier_spec.rb:13:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  2) INotify::Notifier instance run can be stopped before it starts processing
     Failure/Error: run_thread.join(1) or raise "timeout"

     RuntimeError:
       timeout
     # ./spec/notifier_spec.rb:122:in `block in <main>'
     # ./spec/notifier_spec.rb:13:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  3) INotify::Notifier instance to_io returns a ruby IO
     Failure/Error: raise NotImplementedError.new("INotify::Notifier#to_io is not supported under JRuby")

     NotImplementedError:
       INotify::Notifier#to_io is not supported under JRuby
     # ./lib/rb-inotify/notifier.rb:86:in `to_io'
     # ./spec/notifier_spec.rb:134:in `block in <main>'
     # ./spec/notifier_spec.rb:13:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  4) INotify::Notifier instance to_io matches the fd
     Failure/Error: raise NotImplementedError.new("INotify::Notifier#to_io is not supported under JRuby")

     NotImplementedError:
       INotify::Notifier#to_io is not supported under JRuby
     # ./lib/rb-inotify/notifier.rb:86:in `to_io'
     # ./spec/notifier_spec.rb:138:in `block in <main>'
     # ./spec/notifier_spec.rb:13:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  5) INotify::Notifier instance to_io caches its result
     Failure/Error: raise NotImplementedError.new("INotify::Notifier#to_io is not supported under JRuby")

     NotImplementedError:
       INotify::Notifier#to_io is not supported under JRuby
     # ./lib/rb-inotify/notifier.rb:86:in `to_io'
     # ./spec/notifier_spec.rb:142:in `block in <main>'
     # ./spec/notifier_spec.rb:13:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  6) INotify::Notifier instance to_io is selectable
     Failure/Error: raise NotImplementedError.new("INotify::Notifier#to_io is not supported under JRuby")

     NotImplementedError:
       INotify::Notifier#to_io is not supported under JRuby
     # ./lib/rb-inotify/notifier.rb:86:in `to_io'
     # ./spec/notifier_spec.rb:147:in `block in <main>'
     # ./spec/notifier_spec.rb:13:in `block in <main>'
     # ./spec/notifier_spec.rb:8:in `block in <main>'

  7) INotify::Notifier mixed instances doesn't tangle fds
     Failure/Error: raise NotImplementedError.new("INotify::Notifier#to_io is not supported under JRuby")

     NotImplementedError:
       INotify::Notifier#to_io is not supported under JRuby
     # ./lib/rb-inotify/notifier.rb:86:in `to_io'
     # ./spec/notifier_spec.rb:173:in `block in <main>'

Finished in 1.89 seconds (files took 10.52 seconds to load)
14 examples, 7 failures
  </pre>
</details>

**Comment by headius** [comment]
@ahorek So it looks like the change is that `to_io` is now supported and being used under JRuby, which then fails because we use `fstat` in our `IO.new(fileno)` implementation.

**Comment by headius** [comment]
Looping back again...

WSL does not appear to support using `fstat` on this descriptor, but perhaps it will work with `fcntl`?

In CRuby they prefer this when checking the incoming file descriptor for liveness:

```c
#if defined(HAVE_FCNTL) && defined(F_GETFL)
    oflags = fcntl(fd, F_GETFL);
    if (oflags == -1) rb_sys_fail(0);
#else
    if (fstat(fd, &st) < 0) rb_sys_fail(0);
#endif
```

I will try modifying our logic to do the same and see if it helps.

**Comment by headius** [comment]
Ok I have pushed #6535 to address this. I tested it under WSL1 and the example script by @ahorek runs to completion now.

I can't test WSL2 at the moment because I am running MacOS and my system does not support the nested virtualization required to run WSL2 inside virtualized Windows 10.

Anyone able to give that branch a try under WSL2 and let us know if it is working?

**Comment by headius** [comment]
Moving to 9.3 because this is a fairly narrow use case and it will be safer to make this change in 9.3.

**Comment by headius** [comment]
I believe the fixes in #6535 will work around this well enough for now. The basic fix was to ignore EPERM happening during this initial fd-wrapping and assume it means that the file is not a normal file (the main reason we check fstat at this point).
